compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,101 @@ OMR::Node::isNotCollected()
    return false;
    }
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   if (self()->getOpCode().hasPinningArrayPointer())
+      return false;
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collectable.
+ *
+ * \return
+ *    Return true if the node is collectable, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (!curNode->getType().isAddress())
+      return false;
+
+   while (curNode)
```

Comment:
@andrewcraik @jdmpapin Sorry I didn't notice your comments until just now. I still couldn't come up with a case where we can have cycles in the tree, which is the reason causing an exponential walk. And it makes no sense for a tree to have cycles whichever path is taken. The only problem I see here is that the two children of aternary can be the same node. I can check that to avoid examining the same node twice.